src/lib/components/Modal/Modal.jsx

```diff
@@ -78,54 +79,120 @@ export const Modal = ({
+    // to submit the form by pressing Enter key. If there are neither, it tries to focus any other
+    // focusable elements. In case there are none, childrenWrapperElement (Modal itself) is focused.
+
+    if (childrenWrapperRef?.current == null) {
+      return () => {};
     }
-  };
 
-  useEffect(() => {
-    window.document.addEventListener('keydown', keyPressHandler, false);
-    const removeKeyPressHandler = () => {
-      window.document.removeEventListener('keydown', keyPressHandler, false);
-    };
+    const childrenWrapperElement = childrenWrapperRef.current;
+    const childrenFocusableElements = Array.from(
+      childrenWrapperElement.querySelectorAll('button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])'),
+    );
 
-    // If `autoFocus` is set to `true`, following code finds first form field element
-    // (input, textarea or select) or primary button and auto focuses it. This is necessary
-    // to have focus on one of those elements to be able to submit form by pressing Enter key.
-    if (autoFocus) {
-      if (childrenWrapperRef?.current != null) {
-        const childrenWrapperElement = childrenWrapperRef.current;
-        const childrenElements = childrenWrapperElement.querySelectorAll('*');
-        const formFieldEl = Array.from(childrenElements).find(
-          (element) => ['INPUT', 'TEXTAREA', 'SELECT'].includes(element.nodeName) && !element.disabled,
-        );
-
-        if (formFieldEl) {
-          formFieldEl.focus();
-          return removeKeyPressHandler;
-        }
+    const firstFormFieldEl = childrenFocusableElements.find(
+      (element) => ['INPUT', 'TEXTAREA', 'SELECT'].includes(element.nodeName) && !element.disabled,
+    );
+
+    if (childrenFocusableElements.length === 0) {
+      childrenWrapperElement.tabIndex = 1;
+      childrenFocusableElements.push(childrenWrapperElement);
+    }
+
+    const firstFocusableElement = childrenFocusableElements[0];
+    const lastFocusableElement = childrenFocusableElements[childrenFocusableElements.length - 1];
+
+    if (firstFormFieldEl) {
+      firstFormFieldEl.focus();
+    } else if (primaryButtonRef?.current != null) {
+      primaryButtonRef.current.focus({ focusVisible: true });
+    } else {
+      firstFocusableElement.focus();
+    }
+
+    const keyPressHandler = (e) => {
+      if (e.key === 'Escape' && closeButtonRef?.current != null) {
+        closeButtonRef.current.click();
       }
 
-      if (primaryButtonRef?.current != null) {
-        primaryButtonRef.current.focus();
+      if (
+        e.key === 'Enter'
+        && e.target.nodeName !== 'BUTTON'
+        && e.target.nodeName !== 'TEXTAREA'
+        && e.target.nodeName !== 'A'
+        && primaryButtonRef?.current != null
+      ) {
+        primaryButtonRef.current.click();
       }
+
+      // Following code enforces focus to stay inside Modal
+
+      if ((e.key !== 'Tab' && e.keyCode !== 9) || !focusEnforce) {
```

Comment:
I think these key codes (not only here) deserve their constants, at least the numeric one.

Reply:
in the meantime I came across the information that `e.keyCode` is no longer supported in favour of `e.key`. I removed the numeric one. Given that, I don't see much benefit in creating a file with key name constants, since we only use them here and even if we might use them elsewhere in the future, it adds complexity for the sake of complexity and makes the code less readable. In general, I understand the benefits of having it in one place, in this case it seems to me like using a sledgehammer to crack a nut, though. Also, ’Escape’, ’Tab’, ‘Enter’ are self-explanatory enough.

Reply:
As long as there are no magic numbers, I am fine 🙂.